Refactor generic provider search

p0psicles opened the pull request:
This started as a PR to change the following apparent issue:
Apparently someone decided that when finding new results we store them in cache and then continue. This results in users having to perform two forced searches for a single episode.

While trying to come up with a proper fix for this issue. I noticed I had to make some more changes to fix it, without causing any new bugs. That's why i've decided to rewrite part of the GenericProvider search code.

This PR's has made some design choices, that could be debatable. And i'm open for that. I can provide arguments for most of them, but if anything can be done in a better/different way, let me know. In short i've made the following changes:

* Added a new class EvaluateSearchResult(SearchResult), that can be used as a Search result object, before we know if we want to keep this search result. Where the SearchResult was created at the end of "find_search_results" method, this is created at the start, and keeps more information. It could be argued to move all the logic for this class to SearchResult. But I like to keep it separated for now, although it doesn't have much use.
https://github.com/pymedusa/Medusa/pull/1997/files#diff-90b4db298132f61ff3ce451582cd4a63R122

* ASIS evaluated a providers result, for some criteria, like are we searching for a season pack (no), but the result is a season pack, then we dont want it now, but store it for later use. I've changed it to always store it in db. Then we start with a result that we want. And depending on passing some rules, it could change to a result we don't want.

* ASIS had a rule in tv_cache, where it set the season to 1, when it was not known. I've moved this to NameParser(), and only applied this to anime shows.
Note! I've added a warning. We might want to remove this. For now, it provides me better insight.
https://github.com/pymedusa/Medusa/pull/1997/files#diff-d17c57ab2ebb88b02c4ffb79a717869dR210
issue: #2013 

* ASIS had rule that would compare the parsed season (that's what the parser does, it already translates back) with the shows (indexer) season. I've changed to always compare it to the shows season.
@ratoaq2, i've also tried changing NameParser to return the scene_season, but that results in issues in other parts of medusa, where it compares against season.
https://github.com/pymedusa/Medusa/pull/1997/files#diff-9d0f0252087dd8c8f49e52642069a136R290

* ASIS used an any() with some evaluations. I've cut these up into smaller (better readable) evaluations.
https://github.com/pymedusa/Medusa/pull/1997/files#diff-9d0f0252087dd8c8f49e52642069a136L247

* ASIS did a NameParse for every item in the GenericProviders find_search_result method. And then for items that needed to be cached, they where parsed again. Although by then they where probably retrieved from cache, I think this is still resources wasted. I've added the parse result, to the result object. If available tv_cache will also use it.
https://github.com/pymedusa/Medusa/pull/1997/files#diff-cf2820685ee338f116fea75df512b6e2R370

* I've moved some of the result evaluations to the EvaluateSearchResult() class. This improves readability.

## Performance
I have to confess, I do not know what impact these changes have on performance. So we should try to perform some comparable test on low-end hardware using some larger libraries.

- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

p0psicles commented:
@pymedusa/developers, as i'm touching upon a piece of 'old' code here, i'd like to ask if you all can give this a closer look.

p0psicles commented:
Like discussed, the solution is a little bit more complex then I initially thought. The caching is currently too entangled with the evaluation code. So this needs to be more componetized. I'll need some more time to fix this properly.

p0psicles commented:
Still WIP, but now it will cache all items AND keep searching for a result.

fernandog commented:
just tested. working very nice! nice job

p0psicles commented on medusa/providers/generic_provider.py:
```diff
@@ -322,64 +302,47 @@ def __str__(self):
 
-                        if not search_result.add_cache_entry and not search_result.same_day_special:
+                        # @TODO: Need to verify and test this.
+                        if search_result.episode_wanted and not search_result.same_day_special:
```
@duramato or @fernandog anyway you guys know a way of testing this?

duramato replied:
If a special episode airs in the day the search is beeing done?

fernandog commented:
@p0psicles this doesn't happen in develop:

```
2017-01-23 18:40:48 DEBUG    SEARCHQUEUE-DAILY-SEARCH :: [HDTorrents] :: [4f96d96] Traceback (most recent call last):
  File "/home/osmc/Medusa/medusa/search/queue.py", line 252, in run
    found_results = search_for_needed_episodes()
  File "/home/osmc/Medusa/medusa/search/core.py", line 438, in search_for_needed_episodes
    best_result = pick_best_result(cur_found_results[cur_ep], cur_ep.show)
  File "/home/osmc/Medusa/medusa/search/core.py", line 287, in pick_best_result
    cur_result.provider.name):
AttributeError: 'NoneType' object has no attribute 'name'
```

p0psicles commented:
@fernandog can you try to reproduce it with latest commit?
Think I fixed it.
Also I implemented @h3llrais3r's fix. So that should also directly snatch, and add all anime/scene_numbered episodes to cache.

Up until now, I only tested with the more complex anime shows. So I hope I havent's caused any regression.

labrys commented on medusa/classes.py:
```diff
@@ -15,7 +15,7 @@
 # You should have received a copy of the GNU General Public License
 # along with Medusa. If not, see <http://www.gnu.org/licenses/>.
-
+"""Classes module."""
```
These kind of module docstrings are useless.  I'd rather have a flake error than useless documentation, that at least serves to show where improved documentation is needed.  Give some useful information.  For example what kind of classes?  Math? Ballet? :D

p0psicles replied:
Yeah I agree. Problem here is that this is kind of collection of unrelated classes. So the file in it self shouldn't exist imo.

labrys commented on medusa/classes.py:
```diff
@@ -129,15 +130,17 @@ def __init__(self, item, download_current_quality):
 
     def add_result_to_cache(self, cache):
-        # Cache the item if needed
+        """Cache the item if needed."""
```
Same with this docstring.  What type of item. When would it be needed?  A better example would be "Cache a search result if not already cached."

labrys commented on medusa/classes.py:
```diff
@@ -129,15 +130,17 @@ def __init__(self, item, download_current_quality):
 
     def check_episodes_for_quality(self, forced_search, download_current_quality):
-        """We could have gotten a multi-ep result, let's see if at least one if them is wat we want
+        """Check if that episode is wanted in that quality.
```
"Check if an episode is desired at a given quality."

fernandog commented:
@p0psicles 

```
2017-01-24 10:27:51 DEBUG    SEARCHQUEUE-MANUAL-247897 :: [6d946da] 247897: Formatting pattern: %SN - S%0XSE%0XE - %EN -> Homeland - S06E02 - The Man in the Basement
Exception in thread SNATCHQUEUE-MANUALSNATCH-247897:
Traceback (most recent call last):
  File "C:\Python27\lib\threading.py", line 801, in __bootstrap_inner
    self.run()
  File "C:/Users/fernando/Documents/Medusa\medusa\search\queue.py", line 396, in run
    search_result.url = self.cached_result['url']
IndexError: Index must be int or string
```

fernandog commented:
can we merge as it is and continue to improve in another PR? really need to continue on this one?

p0psicles commented:
I'd rather have a review on it first

fernandog commented:
my point is: don't commit more on this PR. let review & merge

p0psicles commented:
I haven't committed more on this Pr. But I also don't see it go anywhere. Any reason to not merge it? Then let me know. Maybe I can improve allot on this.  But with the level of feedback I'm getting. I don't really feel like it.

labrys commented:
Sorry I kinda forgot about this PR... this week is pretty busy for me, but if I get the time, I'll try to review it

p0psicles commented:
Okay guy, i'd like to have this reviewed this week, or merged. Just like labrys doesn't like to constantly have to rebase, I also don't like it. Just had to do a big one, becuase of the refactor to tv_cache -> tv.cache.

fernandog commented:
@p0psicles 

/home/travis/build/pymedusa/Medusa/medusa/tv/cache.py:413:1: D202 No blank lines allowed after function docstring

labrys commented:
@p0psicles If pretty much everything is using `EvaluateSearchResult` is there even a reason to subclass instead of extending the `SearchResult` functionality?  Also a class should generally be named as an object not an action, so it should be `EvaluatedSearchResult` or something similar.

labrys commented on medusa/classes.py:
```diff
@@ -91,6 +91,9 @@ def __init__(self, episodes):
         # Result type like: nzb, nzbdata, torrent
         self.resultType = u''
 
+        # Store the parse result, as it might be usefull of other information later on.
```
"Store the parse result, as it might be ~~usefull~~ useful ~~of~~ for other information later on"

p0psicles replied:
done

labrys commented on medusa/classes.py:
```diff
@@ -129,15 +130,17 @@ def __init__(self, item, download_current_quality):
-        # Cache the item if needed
+        """Cache the item if needed."""
         if self.add_cache_entry:
             logger.debug('Adding item from search to cache: {release_name}', release_name=self.name)
```
This kind of logging is invalid.  Either create and use a brace message adapter or .format the log message or use traditional % formatting.
A braced-message would have a signature as follows:
```python
logger.debug('Adding item from search to cache: {release_name}', {release_name=self.name})
```
Notice the args are a dict.

p0psicles replied:
No this works perfectly. @ratoaq2 added it. I've been using it in numerous of locations.

p0psicles commented:
I was not really sure how Search result was being used in the beginning. So decided to create a subclass. But that's indeed one of the things I was  having doubt about.

p0psicles commented:
I'll consolidate the classes.

fernandog commented on medusa/classes.py:
```diff
@@ -88,54 +89,128 @@ def __init__(self, episodes):
+        self.actual_episodes = None
+
+        # Flag for multi-epp search result.
+        self.multi_epp = False
```
will this be used for a future feature?

p0psicles replied:
I don't know. I think I had a use for it. But can't remember what it was. I'll remove it. As we can already see if it is a mult_epp result, evaluating self.actual_episodes.

fernandog commented on medusa/providers/generic_provider.py:
```diff
@@ -234,139 +234,166 @@ def find_search_results(self, show, episodes, search_mode, forced_search=False,
                                 logger.DEBUG
                             )
-                            add_cache_entry = True
-
```
@p0psicles if you remove this we are not going to cache this found result. We cache so later a user may want a season pack for that show. I didn't understood what is your goal here

p0psicles replied:
It's an opt-out. The add_cache_entry is set to True by default. Then only under rare conditions (mostly exceptions) it's toggled to False.
https://github.com/pymedusa/Medusa/pull/1997/files#diff-90b4db298132f61ff3ce451582cd4a63R108

fernandog replied:
ok, got it

fernandog commented on medusa/providers/generic_provider.py:
```diff
@@ -234,139 +234,166 @@ def find_search_results(self, show, episodes, search_mode, forced_search=False,
+                                "The result %s doesn't seem to have a valid season that we are currently trying to "
+                                "snatch, skipping it" % search_result.name, logger.DEBUG
                             )
-                            add_cache_entry = True
```
also here

p0psicles replied:
same

fernandog replied:
ok!

fernandog commented:
https://github.com/pymedusa/Medusa/issues/2003
